pkg/packet/bgp/bgp.go

```diff
@@ -12325,6 +12360,49 @@ func NewPathAttributeUnknown(flags BGPAttrFlag, typ BGPAttrType, value []byte) *
 	}
 }
 
+// BGPUpdateAttributes defines a map with a key as bgp attribute type
+// and value as bool. Value set to true indicates that the attribute specified by the key
+// exists in the bgp update.
+type BGPUpdateAttributes struct {
+	Attribute map[BGPAttrType]bool
+}
+
+func GetBGPUpdateAttributes(data []byte) map[BGPAttrType]bool {
+	m := make(map[BGPAttrType]bool)
+	for p := 0; p < len(data); {
+		flag := data[p]
+		p++
+		t := data[p]
+		m[BGPAttrType(t)] = true
+		p++
+		var l uint16
+		// Checking for Extened
+		if flag&0x10 == 0x10 {
+			if p+2 <= len(data) {
+				l = binary.BigEndian.Uint16(data[p : p+2])
+			} else {
+				break
+			}
+
+			p += 2
+		} else {
+			l = uint16(data[p])
+			p++
+		}
+		p += int(l)
+	}
+	return m
+}
+
+func GetBGPUpdateAttributesFromMsg(msg *BGPUpdate) map[BGPAttrType]bool {
+	m := make(map[BGPAttrType]bool)
```

Comment:
it is possible to give an allocation hint, like:
```
m := make(map[BGPAttrType]bool, len(msg.PathAttributes))
```

Reply:
I would rather not do it, because we add only sane attributes to the map,  if I preallocate the size based on the length of msg.PathAttributes and during populating the map I will have to skip an attribute by some unexpected reasons I do not want to end up with empty elements. Besides, the number of attributes is relatively small 10 max. I do not see much gain with this change.